storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+	multi := io.TeeReader(defragmented, counter)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerID, ok := s.Layers[blob]; ok {
+					parentLayer = layerID
+				}
+			}
+		}
+		// Now try to figure out if the identifier we have is an ID or
+		// something else, so that we can do collision detection
+		// correctly.
+		names := []string{}
+		id := blobinfo.Digest
+		if matches := idRegexp.FindStringSubmatch(blobinfo.Digest); len(matches) > 1 {
+			// Looks like a digest.  Use the value as a layer ID.
+			id = matches[len(matches)-1]
+		} else {
+			// If it isn't empty, it's just a name.
+			if blobinfo.Digest != "" {
+				names = append(names, blobinfo.Digest)
+				id = ""
+			}
+		}
+		if len(names) == 0 {
+			names = nil
+		}
+		// Attempt to create the identified layer and import its contents.
+		layer, err := s.store.PutLayer(id, parentLayer, names, "", true, multi)
+		if err != nil && err != storage.ErrDuplicateID {
+			logrus.Debugf("error importing layer blob %q: %v", blobinfo.Digest, err)
+			return errorBlobInfo, err
+		}
+		if err == storage.ErrDuplicateID {
+			// We specified an ID, and there's already a layer with
+			// the same ID.  Drain the input so that we can look at
+			// its length and digest.
+			_, err := io.Copy(ioutil.Discard, multi)
+			if err != nil && err != io.EOF {
+				logrus.Debugf("error digesting layer blob %q: %v", blobinfo.Digest, err)
+				return errorBlobInfo, err
+			}
+			hash = hasher.Sum(nil)
+		} else {
+			// Applied the layer, either with a specified ID or a
+			// new ID.  Note the size info and computed digest.
+			hash = hasher.Sum(nil)
+			layerdata := storageLayerMetadata{
+				ExpectedSize: blobinfo.Size,
+				Digest:       "sha256:" + hex.EncodeToString(hash[:]),
+				Size:         counter.Count,
+			}
+			if metadata, err := json.Marshal(layerdata); len(metadata) != 0 && err == nil {
+				s.store.SetMetadata(layer.ID, string(metadata))
+			}
+			// Hang on to the new layer's ID.
+			id = layer.ID
+		}
+		actualSize = counter.Count
+		// If the ID was a digest, verify that our computed sha256sum
+		// matches the ID. */
+		if strings.HasPrefix(blobinfo.Digest, "sha256:") && blobinfo.Digest != "sha256:"+hex.EncodeToString(hash[:]) {
+			logrus.Debugf("blob %q digests to %q, rejecting", blobinfo.Digest, hex.EncodeToString(hash[:]))
+			if layer != nil {
+				// Something's wrong; delete the newly-created layer.
+				s.store.DeleteLayer(layer.ID)
+			}
+			return errorBlobInfo, ErrBlobDigestMismatch
+		}
+		// If we didn't get a name, we might as well assign one using the hash.
+		if digest == "" {
+			digest = "sha256:" + hex.EncodeToString(hash[:])
+		}
+		// Record that this blob is a layer.
+		s.Layers[digest] = id
```

Comment:
Yes, that can happen, but I think only if the random ID the library attempted to assign was a duplicate.  That's something to fix in the storage library.